### PR TITLE
CI: Only trigger on actual code changes

### PR DESCRIPTION
**Pull request by TheRealKeto**
CI builds should only trigger if actual code changes have been made (in any branch or in any PR made to `main`, not `master`) - that is, if a file directly affecting the functionality and building of the project changes.

Also, Github will completely deprecate Node12-based actions (see: [this GH blog post](https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/)), so it's best to upgrade all used actions to their latest versions to prevent the CI from breaking.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -2,15 +2,26 @@ name: CI
 
 on:
```
Comment (saagarjha) [inline .github/workflows/ci.yml:3]: Honestly probably worth merging these entirely:

```suggestion
on: [push, pull_request]
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -27,9 +38,9 @@ jobs:
...
         run: .build/release/unxip Xcode_14.0.1.xip
 
-        # diff on Monterey is gnudiff, which exits with EXIT_TROUBLE (2) if it
-        # finds problems. The Xcode bundle has a ruby symlink that is recursive,
-        # which trips this. So don't rely on the status, but instead use the
-        # output exclusively.
+        # diff on Monterey is gnudiff, which exits with EXIT_TROUBLE (2) if
+        # it finds problems. The Xcode bundle has a Ruby symlink that is
+        # recursive, which trips this. So don't rely on the status, but
+        # instead use the output exclusively.
```
Comment (saagarjha) [inline .github/workflows/ci.yml:44]: Why the change?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (TheRealKeto) [reply .github/workflows/ci.yml:3]: Merging these would trigger the CI on every change of every file on every branch and pull request. This is not the purpose of the PR. Not every commit needs to trigger the CI (e.g a README change)

Comment (saagarjha) [reply .github/workflows/ci.yml:3]: I don't follow? The path filter would still apply, no?

Comment (TheRealKeto) [reply .github/workflows/ci.yml:3]: [It's best to follow Github's example and appropriate YAML syntax.](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-paths)

Comment (saagarjha) [reply .github/workflows/ci.yml:3]: …I'm not sure what you're pointing at?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -2,15 +2,26 @@ name: CI
...
+      - '*.swift'
+      - 'Makefile'
+      - '.github/workflows/*'
+    branches:
+      - main
 
 jobs:
-  test:
+  build:
+    name: Build
```
Comment (saagarjha) [inline .github/workflows/ci.yml:19]: This is named "test" because it runs the binary to make sure it works.

**Review by saagarjha — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (TheRealKeto) [reply .github/workflows/ci.yml:3]: The path filter would not apply with the suggested changes. What I linked is a working example, which mirrors the changes I've introduced which would not work with the suggested changes since it's invalid YAML.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -2,15 +2,26 @@ name: CI
...
+    paths:
+      - '*.swift'
+      - 'Makefile'
+      - '.github/workflows/*'
+    branches:
+      - main
 
 jobs:
-  test:
+  build:
```
Comment (saagarjha) [inline .github/workflows/ci.yml:18]: Why not

```suggestion
  test:
```

**Comment by saagarjha** [comment]
Thanks!
